### ISSUE TITLE
"member ... is inherited ambiguously", "method or attribute is ambiguous", when it isn't

### ISSUE THREAD

**Issue by kingjon3377** [ceylon/ceylon-compiler#2073]
I'm only getting this with objects and classes that extend JMenu; I *suspect* that the fact that the method in question is protected in some superclasses but widened to "public" in a more immediate superclass may be relevant.
```ceylon
import java.awt.event { MouseEvent }
import javax.swing { JMenu }
object mwe extends JMenu("MWE") {
	// With or without (super of) clause (commented below), fails to compile with "method or attribute is ambiguous: 'processMouseEvent' for type 'JMenu'"
	shared actual void processMouseEvent(MouseEvent? mouseEvent) => super.processMouseEvent(mouseEvent);
	//shared actual void processMouseEvent(MouseEvent? mouseEvent) => (super of JMenu).processMouseEvent(mouseEvent);

	// If this line is commented, the class fails to compile with "member 'processKeyEvent' is inherited ambiguously by 'mwe' from 'Component' and another unrelated supertype"
	shared actual void processKeyEvent(KeyEvent? keyEvent) => super.processKeyEvent(keyEvent);
}
```

**Comment by gavinking** [comment]
Confirmed. @FroMage could you take a look, please?

**Comment by tombentley** [comment]
I have something which seems to partially fix this, though I'm not convinced it is right. What seems to go wrong is the `TypeHierarchyVisitor.validateMemberRefinement()` asks the `TypeDeclaration` for `mwe` for its member for `processMouseEvent()`, and it's returning null. It's doing that because, in `Util.lookupMember()`, although a member with that name is found, it's been marked as overloaded due to the changes for #1564 (so although it's marked as overloaded, there is no abstraction method).

So my patch adds an abstraction method in the case that the method is overloaded via an inherited method. 

I say partially because we still can't refine `processMouseEvent()` by delegating to the `super` impl because apparently "method or attribute is ambiguous". And indeed `JMenuItem` has a `processMouseEvent(MouseEvent, MenuElement[], MenuSelectionManager)`. So maybe that error is correct?

But ultimately I guess Gavin's right and @FroMage should take a look.

**Comment by FroMage** [comment]
Oh, sure, dump it on me, as if I remember if we're supposed to have abstract methods for single overloads that overload stuff from the supertype. I do remember that question popped up several times, and unless we already have an issue open to record this, I guess the answer was that we didn't need it.

**Comment by tombentley** [comment]
Well I've pushed my patch as a work in progress, but I'm not sure it's right.

> I guess the answer was that we didn't need it.

Well maybe we didn't *think* we needed it. 

TBH I don't understand how this stuff is supposed to work, though. I mean having a method that's overloaded but there not being an abstraction method *seems* to be wrong, but even when I fix that there are still weird things going on. For example given the Java 

```java
interface Element {
    public void m(String s);
}
class Component {
    public void m(long i){}
}
class Item extends Component implements Element {
    @Override
    public void m(String s){}
}
class Menu extends Item implements Element {
}
```

(which is essentially the same case as `JMenu` in this issue) then the Ceylon:

```ceylon
class C() extends Menu() {
    shared void x(Menu menu) {
        m("");//ok
        super.m("");//ok
        menu.m("");//ok
        
        m(1);//function or value does not exist: 'm'
        super.m(1);//method or attribute is ambiguous: 'm' for type 'Menu'
        menu.m(1);//method or attribute is ambiguous: 'm' for type 'Menu'
    }
}
```

so you can call one method but not the other, and the reason why not isn't even consistent.

Now with a slightly different Java hierarchy:

```java
class X {
    public void m2(String s){}
}
class Y extends X {
    public void m2(long i){}
}
class Z extends Y {
}
```

then the ceylon callsite mostly works, except we can't call one of the methods with a `super` qualifier:

```ceylon
class D() extends Z() {
    shared void x(Z menu) {
        m2("");//ok
        super.m2("");//inherited member is refined by intervening superclass: 'Y' refines 'm2' declared by 'X''
        menu.m2("");//ok
        
        m2(1);//ok
        super.m2(1);//ok
        menu.m2(1);//ok
    }
}
```

Those look like typechecker bugs to me @gavinking, but if they're modelloader bugs you'll have to explain how what we do currently is wrong.

**Comment by FroMage** [comment]
Unless @gavinking responds, Let's move it to 1.3

**Comment by gavinking** [comment]
I don't believe it's a bug in the typechecker unless someone can show me _how_ it is a bug in the typechecker. It sounds like a model loader problem to me.

**Comment by FroMage** [comment]
Well, the thing is that we don't know what the typechecker expects, apparently. I'm moving it to 1.3.

**Comment by FroMage** [comment]
Mmm, it is set to 1.3 already. That issue tracker is nuts. The last move makes it look like _I_ moved it from 1.3 to 1.2 but it's at 1.3 now. No idea.
